examples/routing_service/udp_socket_adapter/src/SocketStreamReader.cxx

```diff
@@ -0,0 +1,184 @@
+
+        /**
+         * Here we notify Routing Service, that there is data available
+         * on the StreamReader, triggering a call to take().
+         */
+        reader_listener_->on_data_available(this);
+    }
+
+    socket_connection_->dispose_discovery_stream(stream_info_);
+}
+
+SocketStreamReader::SocketStreamReader(
+        SocketConnection *connection,
+        const StreamInfo &info,
+        const PropertySet &properties,
+        StreamReaderListener *listener)
+        : stop_thread_(false),
+          stream_info_(info.stream_name(), info.type_info().type_name())
+{
+    socket_connection_ = connection;
+    reader_listener_ = listener;
+    adapter_type_ =
+            static_cast<DynamicType *>(info.type_info().type_representation());
+
+    // Parse the properties provided in the xml configuration file
+    for (const auto &property : properties) {
+        if (property.first == RECEIVE_ADDRESS_STRING) {
+            receive_address_ = property.second;
+        } else if (property.first == RECEIVE_PORT_STRING) {
+            receive_port_ = std::stoi(property.second);
+        } else if (property.first == SHAPE_COLOR_STRING) {
+            shape_color_ = property.second;
+        }
+    }
+
+    // If any of the mandatory properties is not specified, throw exception
+    if (receive_address_.size() == 0 || receive_port_ == 0
+        || shape_color_.size() == 0) {
+        throw dds::core::IllegalOperationError(
+                "You must set receive_address, receive_port and"
+                " shape_color in the RsSocketAdapter.xml file");
+    }
+
+    // Create the UDP socket to receive data
+    socket = std::unique_ptr<UdpSocket>(
+            new UdpSocket(receive_address_.c_str(), receive_port_));
+
+    // Start the receive thread for UDP data
+    socketreader_thread_ =
+            std::thread(&SocketStreamReader::socket_reading_thread, this);
+}
+
+/**
+ * This is the Routing Service take(). It's called when the
+ * socket_receive_thread calls on_data_available()
+ */
+void SocketStreamReader::take(
+        std::vector<dds::core::xtypes::DynamicData *> &samples,
+        std::vector<dds::sub::SampleInfo *> &infos)
+{
+    if (stream_info_.stream_name() == "Square") {
+        /**
+         * This protection is required since take() executes on a different
+         * Routing Service thread.
+         */
+        std::unique_lock<std::mutex> lock(buffer_mutex_);
+        ShapeType *shape = reinterpret_cast<ShapeType *>(received_buffer_);
+        lock.unlock();
+
+        /**
+         * Note that we read one packet at a time from socket_reading_thread()
+         */
+        samples.resize(1);
+        infos.resize(1);
+
+        std::unique_ptr<DynamicData> sample(new DynamicData(*adapter_type_));
+
+        /**
+         * This is the hardcoded type information about ShapeType.
+         * You are advised to change this as per your type definition
+         */
+        sample->value("x", shape->x);
```

Comment:
If we used generated code, I believe we can obtain a dynamic data sample directly from the strong type (there are `convert` functions). As the type complicates (larger types), _that_ would be the way to go, I believe, otherwise it's cumbersome for users.

Reply:
@juanlu-rti please see my comment above. I don't think that using rtiddsgen would help here